### PR TITLE
feat: allow to import custom dashboards

### LINKED ISSUE #63: Superset extensibility points
## Description

Allow to import new dashboards/charts/datasets/databases and override existing dashboards/charts/datasets/databases based on plugins, patches, or custom mechanisms.

## Proposal

Similar to the Cairn implementation but allowing users to patch/override the json or yaml files adapting them to their necessities:

https://github.com/overhangio/tutor-cairn/blob/6f3637e34f798661d88462c0477a7e703f15cf03/tutorcairn/templates/cairn/build/cairn-superset/scripts/cairn#L212-L214

### LINKED ISSUE #75: Superset unable to connect to ClickHouse on fresh Tutor local
As reported in Slack, a new Tutor Local OARS fails to connect to ClickHouse. Changing the connection string from `clickhousedb-connect://...` to `clickhousedb://...` fixes it. It's unclear why this started happening, but we should be able to update the assets to the working connection string.

### PR DESCRIPTION
**Pull request by Ian2012**
## Description

This PR allows importing custom superset assets and overriding default ones by using a tutor patch `superset-extra-assets`.

This has multiple advantages:

- Multiple custom plugins can extend the definition of assets by providing their own assets without breaking anything.
- Developers can override assets by using the same `uuid`.

The script works by creating individual files in their corresponding asset type folder. The patch structure is as follows:

```yaml
name: custom-inline-plugin
version: 0.1.0
patches:
  superset-extra-assets: |
      - file_name: my-dashboard.yaml
        dashboard_title: "..."
        ...
      - file_name: my-chart.yaml
        slice_name: "..."
        ...
      - file_name: my-database.yaml
        database_name: "..."
        ...
      - file_name: my-dataset.yaml
        table_name: "..."
        ...
```

Notice that it's defined as a list, so you can pass multiple assets separated as a **yaml** list item.

##  Screenshots:

image

image

image

image

This PR:
- Fixes #75 
- Fixes #63

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @Ian2012!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

File: tutoroars/plugin.py
Comment (Ian2012) [inline tutoroars/plugin.py:145]: @bmtcril I think we can use a patch for this, defining this large YAML in the **config.yaml** file would be a pain

File: tutoroars/plugin.py
Comment (Ian2012) [inline tutoroars/plugin.py:508]: You can ignore this, it's to speed up testing

**Comment by Ian2012** [comment]
@pomegranited What do you think of this approach?

**Review by mariajgrimaldi — COMMENTED** [review]
Can you explain the changes in the superset jobs/deployments?
https://github.com/openedx/tutor-contrib-oars/pull/69/files#diff-b835bb2b2b3fd535173ffa5f59d95044109d558b0eb20fc22ec72e46759aff56L266-R273
https://github.com/openedx/tutor-contrib-oars/pull/69/files#diff-b835bb2b2b3fd535173ffa5f59d95044109d558b0eb20fc22ec72e46759aff56L266-R273
And so on. I'm a bit confused about what are the changes related to refactoring and what changes actually are adding the new patch 🤔 or adding the new patch include removing those volumes?
Also, can you add what an `asset_definition` would look like? Thanks!

File: tutoroars/templates/oars/jobs/init/superset/superset-api-dashboard.sh
Comment (bmtcril) [inline tutoroars/templates/oars/jobs/init/superset/superset-api-dashboard.sh:20]: Don't suppose we can make this the current timestamp?

**Review by bmtcril — COMMENTED** [review]
Overall this is a huge improvement and looks great. Just some nits!

**Comment by Ian2012** [comment]
@mariajgrimaldi @bmtcril this is ready for review

File: tutoroars/templates/oars/apps/superset/pythonpath/create_assets.py
Comment (mariajgrimaldi) [inline tutoroars/templates/oars/apps/superset/pythonpath/create_assets.py:1]: can we add docstrings here?

**Comment by Ian2012** [comment]
This is currently failing to import the current dashboards due that the passwords can not be sent along with the file database definition, working on it

**Comment by Ian2012** [comment]
@bmtcril @mariajgrimaldi Finally, this is working, please review it now

**Review by bmtcril — APPROVED** [review]
LGTM, I think we should give it until the morning for anyone else to take a look but otherwise I'll merge it!

File: README.rst
Comment (pomegranited) [inline README.rst:96]: I found this a bit confusing..

```suggestion
The patch is expected to be a list of assets with an extra attribute called ``file_name`` , which uniquely identifies the asset entry. This file does not need to exist anywhere; it will be created with the rest of the yaml in that stanza as part of the init process. Each asset is expected to be a valid yaml file with the attributes that superset expects for each asset type. See `assets.yaml <tutoroars/templates/oars/apps/superset/pythonpath/assets.yaml>`_ for examples of asset yaml declarations.
```

**Review by pomegranited — APPROVED** [review]
@Ian2012 Sorry for my delayed reply! This change works great, and using `patches` is a far better way to handle custom import files than what we had before.  👍 

So I don't don't need to block merging here, but I would prefer a mechanism that is easier to use with the zip file exports that Superset produces.

Ideas:

* Instead of having a single `superset-extra-assets` file, is it possible to make that a dir, and import all the *.yaml files in that dir? Then, users could unzip their export file into that dir, and re-init.
* Could we provide a script tool that unzips a superset export zip file and slots all the entries into `superset-extra-assets`?

**Comment by Ian2012** [comment]
@pomegranited What I didn't like about the multiple-folder approach was the complexity of managing that many folders in k8s through config maps, and I didn't find a way to dynamically create files supported by tutor.

On the other hand, there is already a script that converts a zip file and generates a .yaml file with the result.

I will add documentation about it and improve it to be a `tutor do` command

**Comment by bmtcril** [comment]
Looks like some PyCharm files got added by mistake, but you should be good to merge when those last changes are in

**Comment by bmtcril** [comment]
👍 awesome! Feel free to merge once the formatting issues are sorted

**Comment by openedx-webhooks** [comment]
@Ian2012 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/tutor-contrib-oars&entry.867055334=https://github.com/openedx/tutor-contrib-oars/pull/69&entry.1484655318=https://github.com/Ian2012&entry.752974735=2023-06-20+17:23&entry.1917517419=2023-06-22+17:20&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by pomegranited** [comment]
@Ian2012 

> I will add documentation about it and improve it to be a tutor do command

Perfect, thank you!

**Pull request merged**
